js/atree.js

```diff
@@ -1280,6 +1250,62 @@ function atree_set_state(node_wrapper, new_state) {
     }
 };
 
+// first key is connector type, second key is highlight, then [x, y] pair of 0-index positions in the tile atlas
+const atreeConnectorAtlasPositions = {
+    "1100": {"0000": [0, 0], "1100": [1, 0]},
+    "1010": {"0000": [2, 0], "1010": [3, 0]},
+    "0110": {"0000": [4, 0], "0110": [5, 0]},
+    "1001": {"0000": [6, 0], "1001": [7, 0]},
+    "0101": {"0000": [8, 0], "0101": [9, 0]},
+    "0011": {"0000": [10, 0], "0011": [11, 0]},
+    "1101": {"0000": [0, 1], "1101": [1, 1], "1100": [2, 1], "1001": [3, 1], "0101": [4, 1]},
+    "0111": {"0000": [5, 1], "0111": [6, 1], "0110": [7, 1], "0101": [8, 1], "0011": [9, 1]},
+    "1110": {"0000": [0, 2], "1110": [1, 2], "1100": [2, 2], "1010": [3, 2], "0110": [4, 2]},
+    "1011": {"0000": [5, 2], "1011": [6, 2], "1010": [7, 2], "1001": [8, 2], "0011": [9, 2]},
+    "1111": {"0000": [0, 3], "1111": [1, 3], "1110": [2, 3], "1101": [3, 3], "1100": [4, 3], "1011": [5, 3], "1010": [6, 3], "1001": [7, 3], "0111": [8, 3], "0110": [9, 3], "0101": [10, 3], "0011": [11, 3]}
+}
+const atreeConnectorAtlasSize = [17, 5]
+// just has the x position, y is based on state
+const atreeNodeAtlasPositions = {
+    "node_0": 0,
+    "node_1": 1,
+    "node_2": 2,
+    "node_3": 3,
+    "node_archer": 4,
+    "node_warrior": 5,
+    "node_mage": 6,
+    "node_assassin": 7,
+    "node_shaman": 8
+}
+const atreeNodeAtlasSize = [17, 5]
+function atlasBGPositionCalc(pos, atlasSize) {
+    // https://css-tricks.com/focusing-background-image-precise-location-percentages/
+    // p = (c + 0.5/z - 0.5) * z/(z - 1) + 0.5
+    // z = num tiles in direction
+    // c = starting pos of tile in percent of total atlas (equal to index/z)
+    let x = ((pos[0]/atlasSize[0]) + 0.5/atlasSize[0] - 0.5) * atlasSize[0]/(atlasSize[0] - 1) + 0.5
+    let y = ((pos[1]/atlasSize[1]) + 0.5/atlasSize[1] - 0.5) * atlasSize[1]/(atlasSize[1] - 1) + 0.5
+    return (x * 100) + "% " + (y * 100) + "%" // multiply by 100 to convert decimal to percent
+}
+
+// draw the connector onto the screen
+function atree_render_connection(atree_connectors_map) {
+    for (let i of atree_connectors_map.keys()) {
+        let connector_info = atree_connectors_map.get(i);
+        let connector_elem = connector_info.connector;
+        set_connector_type(connector_info);
+        connector_info.highlight = [0, 0, 0, 0];
+        connector_elem.style.backgroundPosition = atlasBGPositionCalc(atreeConnectorAtlasPositions[connector_info.type]["0000"], atreeConnectorAtlasSize);
+        let target_elem = document.getElementById("atree-row-" + i.split(",")[0]).children[i.split(",")[1]];
+        if (target_elem.children.length != 0) {
+            // janky special case...
+            connector_elem.style.display = 'none';
+        }
```

Comment:
oh god

i was gonna suggest changing this but this is disgusting and necessary

nvm!

Reply:
I didn't write this code, I just moved it.

Reply:
yep
and i remember my own stupid, stupid display code